### PR TITLE
refactor: Connect third-party fitness apps

### PR DESCRIPTION
**Pull request by antoncoding**
## TLDR
* Refactor the logic of third-party verifier logins (Strava)
* Change the flow so that an activity is not tied to a specific verification method
* Don't ask users to connect every time
* use `ChallengeType` to replace 

## Check list
- [x] Refactor: Move logic to "Connect Run App" page instead of handling login in `CheckIn`
  - [x] Simplified `CheckIn` page to only check if `verifier` is set, if not, go to `Connect Run App`
  - [x] Cache user's chosen app and the access token locally, so they don't need to reconnect all the time
  - [x]  Abstract away "Strava activity" in the `CheckIn` page, it only use `useRunData` and the branch should happen in the new hook
- [x] Connect Run App Page
  - [x] Connect only changes uri, and redirect to an app-specific callback page. (`connect-run/strava`)
  - [x] Add refresh logic
  - [x] Redirect back to checkin page afterwards
- [x] Remove VerificationType

## Detail
The workflow is now:
 1. Goes into the checkin page, if user has no connected "verifier", ask to go to `/connect-run` page.
 2. User choose which app to connect, redirect to corresponding page after login 
    * strava user will be redirected to `connect-run/strava` page after authorizing on Strava
    * this "Callback" page will bring user back to `CheckIn` page
 3. If user has connected before, but need to refresh access token, this logic is handled in `useRunData`

## Breaking Change
`STRAVA_CLIENT_ID` is now stored as `NEXT_PUBLIC_STRAVA_CLIENT_ID` and can be access by the client
(It will be visible to the client anyway!)

### CODE REVIEW

File: web/next.config.js [hunk 1]
```diff
@@ -14,7 +14,7 @@ const withPWA = require('next-pwa')({
 });
 
 const nextConfig = withPWA({
-  reactStrictMode: true,
+  reactStrictMode: false,
```
Comment (ryanycw) [inline web/next.config.js:17]: Just want to keep a note here, did we encountered any blockers resulting us turning this off?

File: web/app/habit/checkin/components/checkinRun.tsx [hunk 2]
```diff
@@ -4,33 +4,37 @@
 
 import { useState } from 'react';
 import Image from 'next/image';
-import { useParams, useSearchParams } from 'next/navigation';
+import { useRouter } from 'next/navigation';
+import { useParams, usePathname } from 'next/navigation';
```
Comment (ryanycw) [inline web/app/habit/checkin/components/checkinRun.tsx:8]:
```suggestion
import { useParams, usePathname, useRouter } from 'next/navigation';
```

File: web/app/habit/checkin/components/checkinRun.tsx [hunk 3]
```diff
@@ -4,33 +4,37 @@
...
 import { useEffect } from 'react';
 import { useAccount, useWaitForTransactionReceipt } from 'wagmi';
 import { getEncodedCheckinMessage } from '@/utils/arx';
 import toast from 'react-hot-toast';
 import { useWriteContract } from 'wagmi';
 import * as trackerContract from '@/contracts/tracker';
 import { Challenge } from '@/hooks/useUserChallenges';
-import { ActivityTypes } from '@/constants';
-import useStravaData from '@/hooks/useStravaData';
+import { ChallengeTypes } from '@/constants';
```
Comment (ryanycw) [inline web/app/habit/checkin/components/checkinRun.tsx:16]:
```suggestion
```

File: web/app/habit/checkin/components/checkinRun.tsx [hunk 4]
```diff
@@ -4,33 +4,37 @@
...
-export default function StravaCheckIn({ challenge }: { challenge: Challenge }) {
+/**
+ * Running activity check-in page.
+ * @param param0
+ * @returns
+ */
+export default function RunCheckIn({ challenge }: { challenge: Challenge }) {
   const { address } = useAccount();
 
   const { challengeId } = useParams<{ challengeId: string }>();
```
Comment (ryanycw) [inline web/app/habit/checkin/components/checkinRun.tsx:33]:
```suggestion
```

File: web/src/hooks/useRunData.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  }, [expiry, secret, updateVerifierAndSecret, verifier]);
+
+  // fetch data when access token are updated and not expired!
+  useEffect(() => {
+    // TODO: add others verifiers
+    if (verifier !== RunVerifier.Strava) {
+      return;
+    }
+    if (!secret) return;
+    if (expiry < Date.now() / 1000) {
```
Comment (ryanycw) [inline web/src/hooks/useRunData.ts:53]: Just want to keep a note here. I am thinking about if there is any possible way to give the two useEffect in this hook a better readability, the if statements seems a bit confusing in the beginning.

**Review by ryanycw — APPROVED** [review]

File: web/next.config.js (hunk 1 above)
Comment (antoncoding) [reply web/next.config.js:17]: yes, using strict mode cause the `useEffect` hook to run twice, resulting in some weird behaviour when using the authToken to get acessToken
